feat: resolve authentication strategy registered via extension point

emonddr opened the pull request:
Resolve authentication strategies registered via extension point

Part of #2312

## Checklist

👉 [Read and sign the CLA (Contributor License Agreement)](https://cla.strongloop.com/agreements/strongloop/loopback-next) 👈

- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [ ] API Documentation in code was updated
- [x] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [ ] Affected artifact templates in `packages/cli` were updated
- [ ] Affected example projects in `examples/*` were updated

👉 [Check out how to submit a PR](https://loopback.io/doc/en/lb4/submitting_a_pr.html) 👈

emonddr commented:
First commit tackles mock `Basic` authentication strategy class which implements `AuthenticationStrategy` interface, and which registers itself as an extension of the authentication strategy extension point.

jannyHou commented on packages/authentication/src/__tests__/acceptance/authentication-action-provider/authentication-action.provider.ts:
```diff
@@ -0,0 +1,59 @@
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {Getter, inject, Provider, Setter} from '@loopback/context';
```
You can rewrite the [provider](https://github.com/strongloop/loopback-next/blob/master/packages/authentication/src/providers/authentication.provider.ts) directly.

See [the same file in my PR](https://github.com/strongloop/loopback-next/blob/53cfeebbaf2ebe2897e28f51432f6fb5d1a4e946/packages/authentication/src/providers/authentication.provider.ts), which contains exactly the same code as yours, and also proves this action is agnostic of the strategy type(passport or non-passport).

jannyHou commented on packages/authentication/src/__tests__/acceptance/strategy-resolver/strategy.resolver.ts:
```diff
@@ -0,0 +1,45 @@
+    @inject(AuthenticationBindings.METADATA)
+    private metadata: AuthenticationMetadata,
+    @extensions() // Sugar for @inject.getter(filterByTag({extensionPoint: 'greeter'}))
+    private authenticationStrategies: Getter<AuthenticationStrategy[]>,
```
The getter here only captures an array of our authentication strategies, therefore I think we need another extension which has a getter for passport-strategies, like

```ts
@extensionPoint('passport-strategy')
export class PassportStrategyResolverProvider
  implements Provider<Strategy | undefined> {
  constructor(
    @inject(AuthenticationBindings.METADATA)
    private metadata: AuthenticationMetadata,
    @extensions()
    private passportStrategies: Getter<Strategy[]>,
  ) {}
  value(): ValueOrPromise<Strategy | undefined> {
    if (!this.metadata) {
      return;
    }
    // make sure the endpoint is decorated with a passport based strategy
    const isPassportStrategy =
          this.metadata.options && this.metadata.options.isPassportStrategy;
    if (!isPassportStrategy) return;

    const name = this.metadata.strategy;

    return this.findPassportStrategy(name).then(function(strategy) {
      if (strategy) {
        // Please note we wrap it with the adapter to return a Loopback auth strategy
        return new StrategyAdapter(strategy);
      } else {
        throw new Error(`The strategy '${name}' is not available.`);
      }
    });
  }

  async findPassportStrategy(name: string) {
    const strategies = await this.passportStrategies();
    const matchingAuthStrategy = strategies.find(a => a.name === name);
    return matchingAuthStrategy;
  }
}
```

Then the `StrategyResolverProvider` invokes the `PassportStrategyResolverProvider` when it finds the endpoint is decorated with a passport strategy.

jannyHou replied:
You can find the prototype implementation in [My PR](https://github.com/strongloop/loopback-next/blob/53cfeebbaf2ebe2897e28f51432f6fb5d1a4e946/packages/authentication/src/__tests__/acceptance/basic-auth.acceptance.ts#L163-L214).
There are 2 provider classes and each should be decorated with `extensionPoint()`.

emonddr replied:
I will take a look. My strategy resolver in my PR will not deal with passport strategies at the moment. This will be handled by your PR. You can rebase your PR branch off of my branch as discussed.

raymondfeng commented:
Pls fix lint issue - https://travis-ci.org/strongloop/loopback-next/jobs/521377177

jannyHou commented on packages/authentication/src/__tests__/acceptance/keys.ts:
```diff
@@ -0,0 +1,25 @@
+import {BindingKey} from '@loopback/context';
```
Nitpick: I would suggest moving `keys.ts` to a folder called `fixtures`, same for `services` and `strategies`.

jannyHou commented on packages/authentication/src/__tests__/acceptance/users/user.ts:
```diff
@@ -0,0 +1,12 @@
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+export interface User {
```
Usually the `User` would be a model, can we define a model instead?

emonddr replied:
The `user repository` used in similar, existing tests was a hard-coded list of JSON objects and didn't involve datasources and models...so I figured there is no need to create a model for my acceptance tests.

jannyHou commented on packages/authentication/src/types.ts:
```diff
@@ -51,3 +51,54 @@ export interface AuthenticationStrategy {
   authenticate(request: Request): Promise<UserProfile | undefined>;
 }
+
+export interface UserService<U, C> {
```
We created the interface for services in https://github.com/strongloop/loopback-next/blob/master/packages/authentication/src/services/user.service.ts
and 
https://github.com/strongloop/loopback-next/blob/master/packages/authentication/src/services/token.service.ts
No need to duplicate the definition here :-p

emonddr commented on packages/authentication/src/__tests__/acceptance/basic-auth.acceptance.ts:
```diff
@@ -3,207 +3,189 @@
-    cb(null, userList[found].profile);
-  }
-}
+// import {inject, Provider, ValueOrPromise} from '@loopback/context';
```
@jannyHou stated I should comment out these tests for the meantime. Her PR will re-write the test.

jannyHou replied:
@emonddr Could you use `describe.skip()` instead of commenting out the code?
It will be much easier for reviewers to compare the difference in the PR recovering the tests. Thanks!

See the current code change comparison: https://github.com/strongloop/loopback-next/pull/2746/commits/92b10a262233d735d022f752cf4ca82d92f38b09#diff-a5eff729d80d870cc338b162f09467be, all the code are changed, it's hard to see the real difference.

emonddr commented on packages/authentication/src/__tests__/unit/providers/authentication.provider.unit.ts:
```diff
@@ -3,116 +3,116 @@
-import {AuthenticateFn, AuthenticationBindings, UserProfile} from '../../..';
-import {AuthenticateActionProvider} from '../../../providers';
-import {MockStrategy} from '../fixtures/mock-strategy';
+// import {Context, instantiateClass} from '@loopback/context';
```
@jannyHou stated I should comment out these tests for the meantime. Her PR will re-write the test.

emonddr commented:
#2311 will have changes which will affect the `AuthenticationStrategyProvider` presented in this work item. It will involve resolving passport related authentication strategies, and will also affect some mocha tests that I have commented out for the moment.

hacksparrow commented on packages/authentication/docs/authentication-system.md:
```diff
@@ -163,3 +163,192 @@ And the abstractions for:
+`src/providers/auth-strategy.provider.ts` declares an `extension point` named
+`AuthenticationBindings.AUTHENTICATION_STRATEGY_EXTENSION_POINT_NAME` via the
+`@extensionPoint` decorator. `AuthenticationStrategyProvider` is responsible for
+returning an authentication strategy which has a `specific name` and has been
```
Is `specific name` a programming construct? If not, we should omit the backticks. Likewise, other backticked words in the doc.

hacksparrow replied:
@emonddr ^

jannyHou commented on packages/authentication/src/__tests__/fixtures/services/jwt-service.ts:
```diff
@@ -0,0 +1,86 @@
+      );
+    } //if
+
+    if (!userProfile.email) {
```
I remember the `email` and `name` are optional, see [interface UserProfile](https://github.com/strongloop/loopback-next/blob/master/packages/authentication/src/types.ts#L20)

emonddr replied:
@jannyHou:

 I realize that `email` and `name` can be optional in the `UserProfile` interface we have defined in the authentication package, but this is mainly so we don't force the community to use the two fields we are providing as `suggestions`.

 In my acceptance tests  in the `src/__tests__`folder, I create mock implementation classes for two strategies, for a token service, and for a user service, and I personally decided that each field in `UserProfile` (optional or not) is required for creating a JWT Token. The purpose of the mocha tests are to test the authentication strategy interface, and registering authentication strategies via the extensionPoint/extensions pattern.

jannyHou replied:
> and I personally decided that each field in UserProfile (optional or not) is required for creating a JWT Token. 

Fair enough 👍

jannyHou commented on packages/authentication/src/__tests__/fixtures/strategies/jwt-strategy.ts:
```diff
@@ -0,0 +1,47 @@
+  }
+
+  extractCredentals(request: Request): string {
+    if (!request.headers.access_token) {
```
From what I've read online, the typical way is checking the authorization header like `request.header.authorization`.

See "How do JSON Web Tokens work?" in https://jwt.io/introduction/

The `access_token` field is usually from the query, see our implementation in the [shopping cart example](https://github.com/strongloop/loopback4-example-shopping/blob/ee557f0dee0a9a5bd30e1f6b8f63b467d23162ca/packages/shopping/src/authentication-strategies/JWT.strategy.ts#L21).

emonddr replied:
@jannyHou , I changed it to use the Authorization header with 'Bearer aaa.bbb.ccc' (where aaa.bbb.ccc is the JWT token ) instead.

jannyHou commented on packages/authentication/src/types.ts:
```diff
@@ -51,3 +51,13 @@ export interface AuthenticationStrategy {
+ * cannot find the specified authentication strategy by name.
+ */
+export class AuthenticationStrategyNotFoundError extends Error {
+  constructor(error: string) {
```
You may want to add a `code` property as the machine-readable error code that could be understood by any client(HTTP/GRPC/SOAP...)

emonddr replied:
> You may want to add a code property as the machine-readable error code that could be understood by any client(HTTP/GRPC/SOAP...)
@jannyHou , not sure what you mean here.

jannyHou replied:
@emonddr The same topic as what we discussed here https://github.com/strongloop/loopback-next/pull/2763#discussion_r278605367 :) 

```ts
export class AuthenticationStrategyNotFoundError extends Error {
  // the `code` would be a string that could be understood by different clients
  // like 'INVALID_ACCESS_TOKEN'
  // see https://github.com/strongloop/loopback4-example-shopping/blob/master/packages/shopping/src/authentication-strategies/JWT.strategy.ts#L35
  code: string;
  constructor(error: string) { 
    super(error);
  }
}
```

Also see my example https://github.com/strongloop/loopback-next/pull/2746/commits/9bda8b7852db196b72129ea6323635b3b46a7d71#diff-440925ebccbc865e10eb2da8810e08eaR9

emonddr replied:
@jannyHou , I removed the `AuthenticationStrategyNotFoundError` class, and now I just create an instance of Error, and augment it (with Object.assign) with a `code` field  which contains a specific error.

```ts
    return this.findAuthenticationStrategy(name).then(strategy => {
      if (strategy) {
        return strategy;
      } else {
        // important not to throw a non-protocol-specific error here
        let error = new Error(`The strategy '${name}' is not available.`);
        Object.assign(error, {
          code: AUTHENTICATION_STRATEGY_NOT_FOUND,
        });
        throw error;
      }
    });

```

jannyHou replied:
@emonddr thank you, that refactor looks perfect to me 👍

emonddr commented on packages/authentication/src/__tests__/fixtures/keys.ts:
```diff
@@ -0,0 +1,26 @@
+  export const TOKEN_EXPIRES_IN = BindingKey.create<string>(
+    'authentication.jwt.expires.in.seconds',
+  );
+  export const TOKEN_SERVICE = BindingKey.create<JWTService>(
```
Because I am not going to create a binding key with generic interface `UserService`, to be consistent, I won't create a binding key using the generic interface `TokenService`.

emonddr commented on packages/authentication/src/__tests__/fixtures/keys.ts:
```diff
@@ -0,0 +1,26 @@
+);
+
+export namespace BasicAuthenticationStrategyBindings {
+  export const USER_SERVICE = BindingKey.create<BasicAuthenticationUserService>(
```
```ts
  export const USER_SERVICE = BindingKey.create<UserService>(
    'services.authentication.basic.user.service',
  );

```
produces a typescript error most likely because `UserService` is defined like:
```ts
interface UserService<U, C>
```
In my test cases, I only user the `BasicAuthenticationUserService` for my basic authentication strategy mock implementation. I don't see a point in using `UserService` when creating a binding key.

emonddr commented on packages/authentication/src/__tests__/fixtures/services/basic-auth-user-service.ts:
```diff
@@ -0,0 +1,83 @@
+
+  convertToUserProfile(user: User): UserProfile {
+    if (!user) {
+      throw new HttpErrors.Unauthorized(`'user' is null`);
```
I personally like adding `//if` and `//endif` and `//try` and `//catch` next to the `}` in complex nested blocks in order to make my code easier to read...but I will remove the `//if` comments in this piece of code because the code isn't complex.

jannyHou replied:
Oh ok I am ok with either using the comment or not.

raymondfeng replied:
Modern IDE such as VSCode can help find matching `{}` or `()`. Adding such comments is not helpful for other developers as `if` can be nested too.

emonddr commented on packages/authentication/src/__tests__/fixtures/services/basic-auth-user-service.ts:
```diff
@@ -0,0 +1,83 @@
+    credentials: BasicAuthenticationStrategyCredentials,
+  ): Promise<User> {
+    if (!credentials) {
+      throw new HttpErrors.Unauthorized(`'credentials' is null`);
```
@jannyHou , we discussed that the strategy resolver would return a JS error object since, in the future, we may use it to resolve strategies for authenticating protocol requests other than http rest requests, and that the strategy resolver that ships in the main code of the authentication package shouldn't be returning any http specific errors. 

However my mock implementations of token service and user service and strategies don't ship in the main code of the authentication package and only exist in the src/__tests__ folder, and so I have chosen to have then throw http errors since I am testing them with a rest server.

jannyHou replied:
@emonddr Fair enough 👍

emonddr commented:
@jannyHou , I have moved my revised authentication action provider from the main code to the src/tests/fixtures/providers folder since you stated you changed your mind about me replacing the original...that it might cause existing users problems. We will need a new authentication action provider going forward, and the authentication component needs to specify an authentication action provider...so which one will we set? 

I am assuming that your PR will set the appropriate one for the component then.

jannyHou commented:
Was discussing with @emonddr , the next step would be:

- Move the action provider from `__test__/fixtures` to `src/providers`
- Rename the current `authentication-action.provider.ts` to `passport-authentication-action.provider.ts`
- Name the new action provider to be `authentication-action.provider.ts`
- Create a binding key for the deprecated passport auth action provider
- Bind `passport-authentication-action.provider.ts` to the deprecated key
- This PR contains breaking change, let's bump up the version of `@loopback/authentication` to 2.0.0

emonddr commented on packages/authentication/src/__tests__/acceptance/basic-auth.acceptance.ts:
```diff
@@ -3,37 +3,37 @@
 const SequenceActions = RestBindings.SequenceActions;
 
-describe('Basic Authentication', () => {
+describe.skip('Basic Authentication', () => {
```
I only add '.skip(' to the test file, but it also resorted the imports. Not sure how to prevent that.
It shouldn't matter, @jannyHou's PR, https://github.com/strongloop/loopback-next/pull/2746, needs to rewrite this test file.

jannyHou replied:
It's ok, should be some auto fix by prettier/tslint

emonddr commented on packages/authentication/src/__tests__/unit/providers/authentication.provider.unit.ts:
```diff
@@ -108,10 +108,10 @@ describe('AuthenticateActionProvider', () => {
-        () => Promise.resolve(strategy),
-        u => (currentUser = u),
-      );
+      // provider = new AuthenticateActionProvider(
```
Needed to comment out code here. @jannyHou's PR, https://github.com/strongloop/loopback-next/pull/2746, needs to rewrite this test file.

raymondfeng commented on packages/authentication/docs/authentication-system.md:
```diff
@@ -163,3 +163,194 @@ And the abstractions for:
+    }
+    const name = this.metadata.strategy;
+
+    return this.findAuthenticationStrategy(name).then(function(strategy) {
```
Let's consistently use `async-await` instead of `then`.

raymondfeng commented on packages/authentication/src/providers/auth-strategy.provider.ts:
```diff
@@ -0,0 +1,58 @@
+    }
+    const name = this.metadata.strategy;
+
+    return this.findAuthenticationStrategy(name).then(strategy => {
```
Let's use `async-await` instead of `then`.

emonddr replied:
I didn't realize I was allowed to : add `async` to the `value()` function of the `Provider` interface, and change its return type from `ValueOrPromise` to `Promise`.

raymondfeng commented on packages/authentication/docs/authentication-system.md:
```diff
@@ -163,3 +163,194 @@ And the abstractions for:
+
+    this.sequence(SequenceIncludingAuthentication);
+
+    addExtension(
```
Maybe we should add a sugar method such as `app.extension()`.

emonddr replied:
@raymondfeng is this something you will add?

raymondfeng commented on packages/authentication/docs/authentication-system.md:
```diff
@@ -163,3 +163,194 @@ And the abstractions for:
+      BasicAuthenticationStrategy,
+      {
+        namespace:
+          AuthenticationBindings.AUTHENTICATION_STRATEGY_EXTENSION_POINT_NAME,
```
Usually, the namespace is in plural form, such as `authentication.strategies`

raymondfeng commented on packages/authentication/package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@loopback/authentication",
-  "version": "1.2.1",
+  "version": "2.0.0",
```
The `version` should be changed by release process.

raymondfeng commented on packages/authentication/src/__tests__/acceptance/jwt-auth-extension.acceptance.ts:
```diff
@@ -0,0 +1,514 @@
+    server = await app.getServer(RestServer);
+  }
+
+  function givenAuthenticatedSequence() {
```
Can we extract these functions into a separate `helper` ts file so that it can be shared by multiple tests?

emonddr replied:
@raymondfeng , please specify which functions should be extracted and re-used by which tests.   I guess `getAuthenticatedSequence()`  and `givenAServer()` and the `user repo` are common for `packages/authentication/src/__tests__/acceptance/basic-auth-extension.acceptance.ts` and `packages/authentication/src/__tests__/acceptance/jwt-auth-extension.acceptance.ts`.

raymondfeng commented:
@emonddr Quite a few files are missing copyright headers. Please add.

emonddr commented:
added copyright where it was missing.
extracted some common code into a separate sequence class and a helper util.

nabdelgadir commented on packages/authentication/docs/authentication-system.md:
```diff
@@ -163,3 +163,192 @@ And the abstractions for:
+
+  constructor(
+    @inject(BasicAuthenticationStrategyBindings.USER_SERVICE)
+    private user_service: BasicAuthenticationUserService,
```
Question: why aren't you using camel case here?

nabdelgadir commented on packages/authentication/docs/authentication-system.md:
```diff
@@ -163,3 +163,192 @@ And the abstractions for:
+        if (e instanceof AuthenticationStrategyNotFoundError) {
+          throw new HttpErrors.Unauthorized(e.message);
+        } //if
+        else {
```
Nitpick: is the `else` needed?

emonddr replied:
the `else` is needed. If you notice there is a try/catch within a try/catch .

 `await this.authenticateRequest(request);` can fail due to: an error resolving the strategy by name (in a strategy resolver), or due to a strategy throwing an error. 

I want to catch the error  of the `strategy not found` from the resolver (which doesn't throw http errors. it throws a protocol agnostic error), and I've chosen (for my test bucket) to have the sequence throw this error as http error.  

For all other errors (coming from the strategies themselves (they are http errors), I will throw them as-is.

The outer catch catches errors thrown from inner catch.

This is not production code, the sequence handler is a test fixture.

nabdelgadir commented on packages/authentication/src/__tests__/acceptance/jwt-auth-extension.acceptance.ts:
```diff
@@ -0,0 +1,406 @@
+      whoAmI(
+        @inject(AuthenticationBindings.CURRENT_USER) userProfile: UserProfile,
+      ) {
+        if (userProfile) {
```
Suggestion (for here and any other instances of this code):
```ts
if(!userProfile) return 'userProfile is undefined';
if(!userProfile.email) return 'userProfile email is undefined';
return userProfile.email;
```

emonddr replied:
I personally don't like this way of writing if statements.

raymondfeng replied:
What's suggested by @nabdelgadir is consistent with our coding style. The key is to return fast to avoid deep if/else nesting/indentation. It's for readability.

nabdelgadir commented on packages/authentication/src/__tests__/acceptance/jwt-auth-extension.acceptance.ts:
```diff
@@ -0,0 +1,406 @@
+
+    app.controller(InfoController);
+
+    const the_token: string = (await whenIMakeRequestTo(server)
```
Can you just call it `token` (in other tests below too)?

raymondfeng replied:
We should use camelCase for variable names consistently.

nabdelgadir commented on packages/authentication/src/__tests__/acceptance/jwt-auth-extension.acceptance.ts:
```diff
@@ -0,0 +1,406 @@
+describe('JWT Authentication', () => {
+  let app: Application;
+  let server: RestServer;
+  let test_users: UserRepository;
```
Why not `testUsers`?

nabdelgadir commented on packages/authentication/src/__tests__/acceptance/jwt-auth-extension.acceptance.ts:
```diff
@@ -0,0 +1,406 @@
+      .expect(401);
+  });
+
+  it('create a json web token throws error for missing email', async () => {
```
Nitpick (same for the test below): `create` -> `creates`, 
`and throws`?

raymondfeng replied:
We use singular form for the verb to describe a test.

emonddr replied:
@raymondfeng , are you agreeing with me or @nora?

raymondfeng replied:
Yes, I agree with @nabdelgadir to use `creates`.

raymondfeng commented on packages/authentication/src/providers/auth-action.provider.ts:
```diff
@@ -1,14 +1,12 @@
+import {AuthenticateFn, AuthenticationStrategy, UserProfile} from '../types';
 
 /**
  * @description Provider of a function which authenticates
```
Let's remove `@description`.

raymondfeng commented on packages/authentication/src/providers/auth-action.provider.ts:
```diff
@@ -47,12 +45,9 @@ export class AuthenticateActionProvider implements Provider<AuthenticateFn> {
-    return user;
+
+    const userProfile = await strategy.authenticate(request);
+    if (userProfile) this.setCurrentUser(userProfile);
```
Do we need error handling if `userProfile` is not returned from the strategy?

emonddr replied:
the strategy.authenticate(request) implementations will throw an exception when : userService.verifyCredentials(user creds) fails with 401, or when tokenService.verifyToken(token) fails with 401.

emonddr replied:
I will add a check for undefined userProfile.

raymondfeng commented on packages/authentication/src/keys.ts:
```diff
@@ -87,18 +86,22 @@ export namespace AuthenticationBindings {
   );
+
+  export const AUTHENTICATION_STRATEGY_EXTENSION_POINT_NAME =
+    'authentication-strategies';
```
Maybe `authentication.strategies` to be consistent?

raymondfeng commented on packages/authentication/docs/authentication-system.md:
```diff
@@ -163,3 +163,192 @@ And the abstractions for:
+  implements Provider<AuthenticationStrategy | undefined> {
+  constructor(
+    @inject(AuthenticationBindings.METADATA)
+    private metadata: AuthenticationMetadata,
```
private metadata?: AuthenticationMetadata,

raymondfeng commented:
Please either remove https://github.com/strongloop/loopback-next/blob/dremond_authentication_2312/packages/authentication/src/services/session.service.ts or make it an empty interface with copyright headers for now.

raymondfeng commented on packages/authentication/src/services/token.service.ts:
```diff
@@ -1,3 +1,8 @@
+
 import {UserProfile} from '../types';
 
 export interface TokenService {
```
Please add jsdocs.

raymondfeng commented on packages/authentication/src/__tests__/acceptance/basic-auth-extension.acceptance.ts:
```diff
@@ -0,0 +1,199 @@
+    const hash = Buffer.from(credential).toString('base64');
+    await client
+      .get('/whoAmI')
+      .set('Authorization', 'Basic ' + hash)
```
Please refactor the logic to build `Authorization` header from user/pass into a helper function and reuse it for all tests.

emonddr replied:
I will create a utility to base64 encode `email:password` or any combination of strings required for my tests (valid and invalid combinations)

e.g.
```
.set('Authorization', 'Basic ' + createBase64Hash(users.list['joe@example.com'].email, users.list['joe@example.com'].password))
or 
.set('Authorization', 'NotB@s1c ' + createBase64Hash(users.list['joe@example.com'].email, users.list['joe@example.com'].password))
or 
.set('Authorization', 'Basic ' + createBase64Hash(users.list['joe@example.com'].email, users.list['joe@example.com'].password, {alternativeSeparator: '|'}))
or
.set('Authorization', 'Basic ' + createBase64Hash(users.list['joe@example.com'].email, users.list['joe@example.com'].password, {extraSegment: 'extraPart'}))
```

raymondfeng commented on packages/authentication/src/__tests__/acceptance/basic-auth-extension.acceptance.ts:
```diff
@@ -0,0 +1,199 @@
+    app.controller(InfoController);
+    await whenIMakeRequestTo(server)
+      .get('/status')
+      .expect(401);
```
Should we use a different http status code when a strategy is not available?

emonddr replied:
regarding
```ts
it('returns error for unknown authentication strategy', async () => {
    class InfoController {
      @get('/status')
      @authenticate('doesnotexist')
      status() {
        return {running: true};
      }
    }

    app.controller(InfoController);
    await whenIMakeRequestTo(server)
      .get('/status')
      .expect(401);
  });
```
If the user decorates the controller method with the `@authenticate` decorator, then this means he/she wants the endpoint protected. If the user specifies a `name` that he/she hasn't registered as an extension, then I strongly believe the user should receive a `401 - Not Authorized`. If we don't throw this error, this will allow access the endpoint they were trying to protect...this isn't good.

If something goes wrong through any code that is helping with authenticating an endpoint, I am returning a 401 Not authorized in my mock strategy implementations and mock user service and token service.

emonddr replied:
Plus... we don't ship a custom sequence with the authentication action...the user has to create one. They can decide in the sequence code block , what error code to return. It just so happens I decided to return 401...but they can decide otherwise.

raymondfeng commented on packages/authentication/src/__tests__/acceptance/basic-auth-extension.acceptance.ts:
```diff
@@ -0,0 +1,199 @@
+        namespace:
+          AuthenticationBindings.AUTHENTICATION_STRATEGY_EXTENSION_POINT_NAME,
+      },
+    );
```
Let's create a sugar function such as:

```ts
function registerAuthenticationStrategy(strategyClass: Constructor<AuthenticationStrategy>) {
  addExtension(
      server,
      AuthenticationBindings.AUTHENTICATION_STRATEGY_EXTENSION_POINT_NAME,
      strategyClass,
      {
        namespace:
          AuthenticationBindings.AUTHENTICATION_STRATEGY_EXTENSION_POINT_NAME,
      },
    );
}
```

raymondfeng commented on packages/authentication/src/__tests__/acceptance/basic-auth.acceptance.ts:
```diff
@@ -159,32 +156,32 @@ describe('Basic Authentication', () => {
+    // }
+    // server
+    //   .bind(AuthenticationBindings.STRATEGY)
+    //   .toProvider(MyPassportStrategyProvider);
```
It's probably better to use block comment here.

raymondfeng commented on packages/authentication/src/__tests__/acceptance/jwt-auth-extension.acceptance.ts:
```diff
@@ -0,0 +1,406 @@
+        public users: UserRepository,
+      ) {}
+
+      @get('/login')
```
login is usually over http `post`.

raymondfeng commented on packages/authentication/src/__tests__/acceptance/jwt-auth-extension.acceptance.ts:
```diff
@@ -0,0 +1,406 @@
+      .get('/login')
+      .expect(200)).text;
+
+    expect(the_token !== null).to.equal(true);
```
expect(the_token).to.be.not.null();

emonddr replied:
I tried to use something similar a few weeks ago, and a linter didn't like it. This one worked though.

raymondfeng commented on packages/authentication/src/__tests__/acceptance/jwt-auth-extension.acceptance.ts:
```diff
@@ -0,0 +1,406 @@
+      .expect(200)).text;
+
+    expect(the_token !== null).to.equal(true);
+    expect(typeof the_token === 'string').to.equal(true);
```
expect(the_token).to.be.string();

raymondfeng commented on packages/authentication/src/__tests__/fixtures/users/user.repository.ts:
```diff
@@ -0,0 +1,22 @@
+import {User} from './user';
+
+export class UserRepository {
+  constructor(readonly list: {[key: string]: {user: User}}) {}
```
Let's get rid of `user` property:
```ts
constructor(readonly list: {[key: string]: User})
```

raymondfeng commented on packages/authentication/src/__tests__/fixtures/users/user.repository.ts:
```diff
@@ -0,0 +1,22 @@
+    const found = Object.keys(userList).find(search);
+    if (found) {
+      return userList[found].user;
+    } //if
```
No need to have `// if`.

raymondfeng commented on packages/authentication/src/__tests__/fixtures/users/user.repository.ts:
```diff
@@ -0,0 +1,22 @@
+    const userList = this.list;
+    function search(key: string) {
+      return userList[key].user.email === email;
+    }
```
Use an arrow function instead:
```ts
Object.keys(userList).find(k => this.list[k].email === email);
```

b-admike commented on packages/authentication/docs/authentication-system.md:
```diff
@@ -163,3 +163,192 @@ And the abstractions for:
   - controller function:
     - process the injected user
+
+## Registering an authentication strategy via an extension point
```
I hope in follow-up PRs we can move this awesome doc to the README.

emonddr replied:
@b-admike , yes. @jannyHou 's PR : https://github.com/strongloop/loopback-next/pull/2822 overlaps with mine, and we have discussed that we will be moving all separate .md files in the authentication package into a comprehensive readme.md file. And this readme file will also be the source of the loopback.io docs that are generated by our doc build process.

b-admike commented on packages/authentication/src/__tests__/acceptance/basic-auth-extension.acceptance.ts:
```diff
@@ -0,0 +1,199 @@
+      .expect(users.list['joe@example.com'].user.email);
+  });
+
+  it(`authenticates successfully for correct credentials for user 'jill'`, async () => {
```
Feel free to ignore, but I think it'd be good to re-assign `credential` to user `jill` in the test above and assert that auth works for both users (have it in one test)

emonddr replied:
After speaking with @b-admike , we decided to remove the 2nd test involving the user Jill.

b-admike commented on packages/authentication/src/__tests__/acceptance/jwt-auth-extension.acceptance.ts:
```diff
@@ -0,0 +1,406 @@
+  });
+
+  it(`returns error for invalid 'Bearer ' portion of Authorization header value`, async () => {
+    class InfoController {
```
Consider moving `InfoController` to the bottom of the test as a helper if it can be re-used throughout these tests. Feel free to do it in a follow up PR.

emonddr replied:
The Info controllers are different for pretty much all tests, so I will leave as is.

b-admike commented on packages/authentication/src/__tests__/acceptance/jwt-auth-extension.acceptance.ts:
```diff
@@ -0,0 +1,406 @@
+
+    await whenIMakeRequestTo(server)
+      .get('/createtoken')
+      .expect(401);
```
For a lot of these tests we are asserting the error code, but is it possible to also assert error messages sent back? Not a deal breaker, but it'll make the tests more robust.

nabdelgadir replied:
+1

emonddr replied:
@b-admike , I have addressed your comment. I check error code and message in my tests.

b-admike commented on packages/authentication/src/__tests__/fixtures/strategies/jwt-strategy.ts:
```diff
@@ -0,0 +1,50 @@
+    } //if
+
+    //split the string into 2 parts : 'Bearer ' and the `xxx.yyy.zzz`
+    let parts = auth_header_value.split(' ');
```
Do we want to assert that length of parts does not exceed 2 here?

emonddr replied:
I will add a check, but we must remember these are mock implementations in src/__tests__ and not production code. We don't need to test each line of code for these items. This PR has mocha tests that focus on verifying the authentication strategy interface , registering and finding of authentication strategies via extensionPoints, the authentication action, and the authentication strategy provider. The mock implementations for token service and user service and basic strategy and jwt strategy only serve as fixture helpers and don't need to be tested so heavily.

b-admike commented on packages/authentication/src/providers/auth-strategy.provider.ts:
```diff
@@ -0,0 +1,50 @@
+  AuthenticationBindings.AUTHENTICATION_STRATEGY_EXTENSION_POINT_NAME,
+  {scope: BindingScope.TRANSIENT},
+)
+export class AuthenticationStrategyProvider
```
Since this is part of the source code, it'd be great to add TSDocs to here for our API docs.

emonddr commented:
> Please either remove https://github.com/strongloop/loopback-next/blob/dremond_authentication_2312/packages/authentication/src/services/session.service.ts or make it an empty interface with copyright headers for now.

I will make it an empty interface with copyright headers for now.

dhmlau commented:
@emonddr , FYI,  you can install `strong-tools` globally and run `slt copyright` to add the copyright headers.

emonddr commented:
added two more test cases for auth-action.provider.ts since code coverage decreased after I added an error for undefined userProfile return from a strategy.

emonddr commented:
rebased...

raymondfeng commented on packages/authentication/src/__tests__/fixtures/helper.ts:
```diff
@@ -0,0 +1,99 @@
+  // extraPart is '' (blank), unless another is specified
+  const extraPart =
+    options && options.extraSegment ? separator + options.extraSegment : '';
+  return (
```
This can be simplified as:
```ts
options = Object.assign({
  alternativePrefix: 'Basic ',
  alternativeSeparator: ':',
  extraSegment: ''
}, options);
```

emonddr replied:
That's a great replacement for the ternary operator in this case. Thanks. :)

raymondfeng replied:
Thanks for the update. Next time, please click on `Resolve conversation` if it's fixed.

raymondfeng commented on packages/authentication/src/__tests__/fixtures/sequences/authentication.sequence.ts:
```diff
@@ -0,0 +1,80 @@
+          e.code === USER_PROFILE_NOT_RETURNED
+        ) {
+          throw new HttpErrors.Unauthorized(e.message);
+        } //if
```
Please remove `//if`.

emonddr replied:
removed all `//if` occurrences.

raymondfeng commented on packages/authentication/src/__tests__/fixtures/sequences/authentication.sequence.ts:
```diff
@@ -0,0 +1,80 @@
+          // strategy error
+          throw e;
+        } //endif
+      } //catch
```
Please remove `//endif` and `//catch`.

raymondfeng commented on packages/authentication/src/__tests__/fixtures/services/basic-auth-user-service.ts:
```diff
@@ -0,0 +1,80 @@
+      throw new HttpErrors.Unauthorized(`'credentials' is null`);
+    }
+
+    if (!credentials.email) {
```
Basic auth allows `username`. Do we want generalize it as `username` instead of `email`?

emonddr replied:
can't people log in with email and password?

emonddr replied:
email can be a person's userid can it not?

emonddr replied:
I can have two gmail addresses associated with my name and surname. So an email can serve as a username.

emonddr replied:
I am trying to make use of
```ts
export interface UserProfile {
  id: string;
  name?: string;
  email?: string;
}
```
and in our user service interface, we have 
```ts
convertToUserProfile(user: U): UserProfile;
```

where we have a `user` object (we obtained by searching a database with user credentials) and then we convert the larger `user` object into a slimmer `user profile` we place on the context.

Why would we have the email field in UserProfile if we cannot use it?
@jannyHou , what are your thoughts? thx

dhmlau replied:
Not sure how relevant it is, just wanna point out that there is another discussion thread talking about making the UserProfile more flexible: https://github.com/strongloop/loopback-next/issues/2246.

raymondfeng replied:
I'm proposing to be consistent with http://www.passportjs.org/docs/basic-digest/. Please `email` id can be used as `username`.

raymondfeng commented on packages/authentication/src/__tests__/fixtures/services/basic-auth-user-service.ts:
```diff
@@ -0,0 +1,80 @@
+    if (!user.email) {
+      throw new HttpErrors.Unauthorized(`'user.email' is null`);
+    }
+
```
Do we really want to make `firstname`, `surname`, `email` required? It is desired sometimes to hide fields such as `email` due to privacy.

I'm also not sure if these errors should be mapped to `HttpErrors.Unauthorized`.

emonddr replied:
The token service and user service and strategies in src/__tests__/fixtures are mock implementations not production implementations in 'src'. Is it absolutely necessary for them to be perfect? The users must create their own production ready artifacts. (Just wondering)

emonddr replied:
The code block above is called when the basic strategy is trying to verify that the user's credentials are good.

called from 
```ts
 async authenticate(request: Request): Promise<UserProfile | undefined> {
    const credentials: BasicAuthenticationStrategyCredentials = this.extractCredentals(
      request,
    );
    const user = await this.userService.verifyCredentials(credentials);
    const userProfile = this.userService.convertToUserProfile(user);

    return userProfile;
  }

```

raymondfeng replied:
In this case, some of the user profile fields are not populated. It's not something wrong with the credentials. For basic auth, credentials are username/password.

raymondfeng commented on packages/authentication/src/__tests__/fixtures/services/jwt-service.ts:
```diff
@@ -0,0 +1,88 @@
+      // decode user profile from token
+      userProfile = await verifyAsync(token, this.jwt_secret);
+    } catch (error) {
+      throw new HttpErrors['Unauthorized'](
```
Let's use `new HttpErrors.Unauthorized` instead.

raymondfeng commented on packages/authentication/src/__tests__/fixtures/strategies/jwt-strategy.ts:
```diff
@@ -0,0 +1,54 @@
+
+  constructor(
+    @inject(JWTAuthenticationStrategyBindings.TOKEN_SERVICE)
+    public token_service: JWTService,
```
Please use camelCase `tokenService`.

raymondfeng commented on packages/authentication/src/__tests__/fixtures/strategies/jwt-strategy.ts:
```diff
@@ -0,0 +1,54 @@
+    if (!request.headers.authorization) {
+      //throw an error
+      throw new HttpErrors.Unauthorized(`Authorization header not found.`);
+    } //if
```
Please remove `//if`

raymondfeng commented on packages/authentication/src/__tests__/fixtures/users/user.repository.ts:
```diff
@@ -0,0 +1,19 @@
+    if (found) {
+      return this.list[found];
+    }
+    return undefined;
```
It can be simplified as `return found ? this.list[found]: undefined;`.

raymondfeng commented on packages/authentication/src/__tests__/fixtures/strategies/basic-strategy.ts:
```diff
@@ -36,20 +36,20 @@ export class BasicAuthenticationStrategy implements AuthenticationStrategy {
     if (!request.headers.authorization) {
       //throw an error
```
This comment does not add any value here.

bajtos commented on packages/authentication/docs/authentication-system.md:
```diff
@@ -163,3 +163,194 @@ And the abstractions for:
+          e.code === AUTHENTICATION_STRATEGY_NOT_FOUND ||
+          e.code === USER_PROFILE_NOT_RETURNED
+        ) {
+          throw new HttpErrors.Unauthorized(e.message);
```
I feel this is adding too much friction for application developers. The current proposal is is good enough for now (incremental baby steps FTW!), but I'd like you to create a follow-up SPIKE issue to look into ways how to allow  `@loopback/authentication` (and any other 3rd party extensions) to contribute custom error-code to status-code mapping.  See the current hard-coded map here:
https://github.com/strongloop/loopback-next/blob/22400fe9a40b17042a5df684300f8a007dd50edd/packages/rest/src/providers/reject.provider.ts#L12-L16

emonddr replied:
@bajtos :
> I'd like you to create a follow-up SPIKE issue to look into ways how to allow @loopback/authentication (and any other 3rd party extensions) 

@jannyHou  mentioned there was a task to work on something like this.

Do you mean [Enable custom reject implementation to leverage the built-in "code to statusCode" mapping ](https://github.com/strongloop/loopback-next/issues/1942) ?

emonddr replied:
@bajtos 
> I feel this is adding too much friction for application developers

I am not sure what you mean. The code you mention is inside a `custom sequence` that user must write... we are not shipping the sequence which includes the authentication action. The two agnostic errors I am specifying here are thrown from code that cannot throw protocol specific errors since that code may support protocols other than http in the future.  My mock (not production) implementations for strategies, token service and user service need to throw some kind of errors. I decided for my test bucket to have them throw http errors because my test bucket runs a rest server that handles http requests. The custom sequence rethrows these http exceptions as is, and throws the captured agnostic errors as http errors .

Since users will be writing their own production-ready implementations for `sequence`, `token service`, `user service`, and `strategies` , I am having a hard time understanding how the code above is `painting the user into a corner`. If you could clarify. thanks :)

bajtos replied:
@emonddr 

> Do you mean Enable custom reject implementation to leverage the built-in "code to statusCode" mapping ?

Not really. The goal of https://github.com/strongloop/loopback-next/issues/1942 is to allow custom implementations of `reject` to read data from the mapping table. I.e. if the application developer decides to write a different error handling routine, they should have a way how to determine the HTTP status code to return based on error code.

In this discussion, I am asking for a mechanism that will allow extension developers to contribute new mapping entries, so that `reject` implementations can pick them up.

> I am not sure what you mean. The code you mention is inside a custom sequence that user must write...

I find the custom sequence you are proposing as too complex. I would like the custom sequence to look like this:

```ts
class SequenceIncludingAuthentication implements SequenceHandler {
  // ...
  async handle(context: RequestContext) {
    try {
      const {request, response} = context;
      const route = this.findRoute(request);
      
      await this.authenticateRequest(request);	

      const args = await this.parseParams(request, route);
      const result = await this.invoke(route, args);
      this.send(response, result);
    } catch (error) {
      this.reject(context, error);
      return;
    }
  }
}
```

Under the hood:
- When `@loopback/authentication` is registered as a component via `app.component`, it contributes new mappings to `codeToStatusCodeMap`.
 - The built-in `reject` action picks up the mapping contributed by the authentication extension and converts `AUTHENTICATION_STRATEGY_NOT_FOUND` to `401 Unauthorized`.

---

BTW the current implementation is problematic because it discards the original error stack trace pointing to the place where the error was triggered (e.g. in the authentication strategy) and replaces it with a stack trace pointing to the custom sequence. This makes troubleshooting rather difficult.

Here is a better solution:

```ts
try {
  await this.authenticateRequest(request);
} catch (e) {
  if (
    e.code === AUTHENTICATION_STRATEGY_NOT_FOUND ||
    e.code === USER_PROFILE_NOT_FOUND
  ) {
    err.statusCode = 401;
  }
  throw err;
}
```

emonddr replied:
Good point about the stack trace. I will look into it. :)

bajtos commented on packages/authentication/docs/authentication-system.md:
```diff
@@ -163,3 +163,194 @@ And the abstractions for:
+        // strategy not found error, or user profile undefined
+        if (
+          e.code === AUTHENTICATION_STRATEGY_NOT_FOUND ||
+          e.code === USER_PROFILE_NOT_RETURNED
```
I find the code `USER_PROFILE_NOT_RETURNED` not very clear - what does it mean that user profile was not returned? Did the code looking it up freeze and never return back to the caller?

I am proposing to use `USER_PROFILE_NOT_FOUND` code instead, following the convention used by `AUTHENTICATION_STRATEGY_NOT_FOUND` and `ENTITY_NOT_FOUND` (from `@loopback/repository`).
